content/blog/dont-use-preludes-and-globs/index.md

```diff
@@ -0,0 +1,152 @@
+
+I have a hard time reviewing your code which uses preludes: looking up a definition might not work in my code review tool, and I'd have to manually track down where a types came from.
+
+Explicit imports bring clarity.
+
+Speaking of which! A good compromise between convenience and clarity is to import whole modules instead of individual items, such as `use std::fmt`. Then you can say `fmt::Debug` instead of `std::fmt::Debug`.
+
+Unless you write a highly critical framework which provides well-established types, which are *absolutely positively always* mandatory in every (or at least most) interactions with your library, don't add a prelude. Even then, make sure that prelude carries its own weight. [Tokio removed their prelude because that wasn't the case](https://github.com/tokio-rs/tokio/issues/3257).
+
+An explicit import might not be that bad. In fact, it might even clear things up for your users, because it's easier to see where imports come from.
+
+If you're worried about ergonomics, why not re-export common types in root? Then they can get imported with `use mycrate::SomeType;` instead of `use mycrate::mymodule::SomeType;`.
+
+```rust
+// Re-export a type in the your lib.rs
+pub use mymodule::SomeType;
+```
+
+On the same note, avoid glob imports (like `use prelude::*`
+or `use mycrate::mymodule::*`).
+The reasons are the same: they hide dependencies and make it harder to grep for usages of a type or function.
+
+## Common Arguments for Preludes And Glob Imports
+
+Let's address some common arguments in favor of preludes and glob imports
+and see if there's a better way to achieve the same goal.
+
+### Preludes in examples
+
+Some people like to use preludes in their documentation examples.
+
+This can be sweet when you quickly want to show how to use your library without too much boilerplate. But at what cost? At what terrible cost?[^2]
+
+[^2]: I'm exaggerating. Bear with me.
+
+Users *can and will* start to copy-paste that prelude into their code, at which point you'll have to support it. It's convenient in the beginning, but can result in a maintenance burden in the long run.
+
+Instead, why not use hidden code lines for imports in your documentation examples? This way, the examples are still easy to read, but users can't copy-paste the imports:
+
+```rust
+//! # Examples
+//!
+//! ```
+//! # use mycrate::SomeType;
+//! # fn main() {
+//! let t = SomeType::new(); 
+//! # }
+//! ```
+```
+
+This will render as:
+
+```rust
+let t = SomeType::new();
+```
+
+### Reducing boilerplate
+
+Another common argument for preludes is that they reduce friction when importing types.
+
+I don't think that's true anymore. IDE support for Rust has made great strides in the last years. Nowadays, your IDE will just auto-import the type for you. (Largely thanks to [Rust Analyzer](https://rust-analyzer.github.io/) and [Rust Rover](https://www.jetbrains.com/rust/))
+Problem solved!
```

Comment:
```suggestion
I don't think that's true anymore. Editor support for Rust has made great strides in the last years. Nowadays, your editor will just auto-import the type for you. (Largely thanks to [Rust Analyzer](https://rust-analyzer.github.io/) and [Rust Rover](https://www.jetbrains.com/rust/)). Problem solved!
```

I replaced "IDE" with "editor". VS Code barely is an IDE, Rust Rover is. But VIM / Sublime / Zed aren't. And even those do that with LSPs...

And punctuation.